server/user.js

```diff
@@ -0,0 +1,330 @@
+ */
+user.query.signUp = function(query, callback) {
+  var userInfo = {
+    email: query.email,
+    username: query.username,
+    password: CryptoJS.SHA256(query.password).toString(),
+    confirmed: query.confirmed
+  };
+  if (!user.validateUserInfo(userInfo.username, userInfo.password,
+      userInfo.email)) {
+    return;
+  }
+  user.signUp(userInfo, function(result) {
+    callback(result);
+  });
+};
+
+/**
+ * @param {*|!user.Cookie} query
+ * @param {function((!user.Error|!user.Cookie))} callback Callback function.
+ */
+user.query.autoSignIn = function(query, callback) {
+  var cookie = {
+    username: query.username,
+    sessionId: query.sessionId,
+    expiration: query.expiration
+  };
+  user.autoSignIn(cookie, function(result) {
+    callback(result);
+  });
+};
+
+/**
+ * Checks the user information for signing in.
+ * @param {!user.Info} userInfo User Information.
+ * @param {function((!user.Error|!user.Cookie))} callback Callback function.
+ */
+user.signUp = function(userInfo, callback) {
+  var db = database.db;
+  var cursor = db.collection(user.userInfoDbName).find({
+    $or: [{username: userInfo.username}, {email: userInfo.email}]
+  });
+  var documents = [];
+  cursor.each(function(err, doc) {
+    if (err) {
+      log.serverLog(err.message);
+      return;
+    }
+    if (doc != null) {
+      delete doc['_id'];
+      documents.push(doc);
+    } else {
+      var data = user.signUpUser_(documents, userInfo);
+      if (!data) {
+        var cookie = {
+          username: userInfo.username
+        };
+        data = user.updateSession(cookie);
+      }
+      callback(data);
+    }
+  });
+};
+
+/**
+ * Checks the user information for signing in.
+ * @param {!user.Info} userInfo User Information.
+ * @param {function((!user.Error|!user.Cookie))} callback Callback function.
+ */
+user.signIn = function(userInfo, callback) {
+  var db = database.db;
+  var query = {$and: [{username: userInfo.username},
+    {password: userInfo.password}]};
+  database.getOne(db.collection(user.userInfoDbName), query, [],
+    function(result) {
+      var data;
+      if (!result) {
+        data = {
+          error: 'incorrect username or password'
+        };
+      } else {
+        var cookie = {
+          username: userInfo.username
+        };
+        data = user.updateSession(cookie);
+      }
+      callback(data);
+    });
+};
+
+/**
+ * Checks the user session information for auto signing in.
+ * @param {!user.Cookie} cookie User cookie information.
+ * @param {function((!user.Error|!user.Cookie))} callback Callback function.
+ */
+user.autoSignIn = function(cookie, callback) {
+  var db = database.db;
+  var query = {$and: [{username: cookie.username},
+    {sessionId: cookie.sessionId},
+    {expiration: {$gt: new Date().getTime()}}]};
+  database.getOne(db.collection(user.sessionDbName), query, [],
+    function(result) {
+      var data;
+      if (!result) {
```

Comment:
Same for the aborting on error style. Usually we don't write "else" for "error conditioning".
```js
if (error) {
} else { //bad
}

if (error) { // good
  //...
  return;
}
//...
```